### PR TITLE
Simple ambientOcclusion effect

### PR DESCRIPTION
**Pull request by vgerber**
Added ambient occlusion as post processing filter
- Editable params in web interface
- AO for particles (+blending with volume)

Here are some example renderings
ao_aftereffect1
ao_aftereffect2
ao_aftereffect3
ao_comp_landscape

### CODE REVIEW

**Comment by n01r** [comment]
Nice work, awesome! :sparkles: 
So you think you got the GPU edge issue fixed as well?
Can you show the results of a runtime test?

**Comment by vgerber** [comment]
the gpu edge issue is still there. 
I have some ideas for fixing/hiding this issue in the post processing step and will take a look next week.
-  extra checks for border pixels added a huge performance drop

At the moment i just fixed the blending bug with volumes and tried to release a possible usable implementation

**Comment by n01r** [comment]
> ... release a possible usable implementation

This is definitely the correct way. If the possible fix makes the performance drop steeply it might be a good idea to make it optional. 
How large is the difference in performance?
We might want it for really pretty production runs and leave it out for ourselves where we just want to watch our own smaller simulations to understand the dynamics better.

**Comment by vgerber** [comment]
I will test how to minimize the performance drop and add a flag for compiling.
In earlier versions without the blending i only had +-2ms of extra time, but in my latest version with the blending the extra cost went up to 30-40ms so there must be a major flaw in my code (i hope so)

**Comment by n01r** [comment]
Ah, also just for reference: would you mind adding pictures of the bug and how it looks after the fix?
You probably have them lying around, anyway, right?

**Comment by vgerber** [comment]
Well the performance "issue" is back to normal so +-2ms. I completely forgot the other sources were still on and they introduced the additional (30-40ms) computation cost. Therefore i will not add the compile flag because whether you turn the effect on or off you will not notice a performance change.

**Comment by vgerber** [comment]
> Ah, also just for reference: would you mind adding pictures of the bug and how it looks after the fix?
> You probably have them lying around, anyway, right?

Here is an example for the gpu border issue
gpu_border_bug

File: client/js/isaac/interface/visualisation.js [hunk 1]
```diff
@@ -1115,4 +1117,37 @@ function updateWireframe() {
...
+	
+	if (response.hasOwnProperty("ao isEnabled")) {
+		console.log(response["ao weight"]);
+		document.getElementById("ao_checkbox").checked = response["ao isEnabled"];
+		lastIsaacState["ao isEnabled"] = response["ao isEnabled"];
+	}
+	
+	if (response.hasOwnProperty("ao weight")) {
+		document.getElementById("ao_weight").value = response["ao weight"];
+		lastIsaacState["ao_weight"] = response["ao weight"];;
```
Comment (ax3l) [inline client/js/isaac/interface/visualisation.js]:
```suggestion
		lastIsaacState["ao_weight"] = response["ao weight"];
```

File: client/js/isaac/interface/visualisation.js [hunk 2]
```diff
@@ -1115,4 +1117,37 @@ function updateWireframe() {
...
 	renderer.setWireframe(enabled, color);
+}
+
+/**
+ * 
+ * @param {IsaacResponse} response 
+ */
+function aoSetValues(response) {
+	
+	if (response.hasOwnProperty("ao isEnabled")) {
```
Comment (ax3l) [inline client/js/isaac/interface/visualisation.js]: just wondering: are spaces in that key a good idea?

Comment (vgerber) [reply client/js/isaac/interface/visualisation.js]: Typically you would use ao as extra class and isEnabled and weight as it's attributes 
`{
    "ao": {
             "isEnabled": true, 
             ...
             }
}`
but i was unable create extra classes in janson (they were never attached) so i kept this style and would change this if the json overhaul with a more c++ like lib begins (there are some other attributes that have white spaces and are awful to deal with in js like response["test 123"] instead of response.test_123)

So basically yes its a bad style and should not be in the codebase.

Comment (psychocoderHPC) [reply client/js/isaac/interface/visualisation.js]: Yes I also vote for using a modern json API.

**Comment by PrometheusPi** [comment]
Great job! Excellent that AO is now available in the mainline. :+1:
